### PR TITLE
Resolve postponed annotations in handler validation

### PR DESCRIPTION
**Pull request by moonbox3**
## Summary
- resolve handler annotations via `get_type_hints` with fallback string resolution before WorkflowContext validation
- reuse resolved ctx annotation for explicit handler mode
- add guard for unresolved string WorkflowContext annotations
- add regression coverage for __future__ annotations and forward refs in WorkflowContext generics

## Testing
- `python/packages/core/tests/workflow/test_handler_future_annotations.py::test_handler_future_annotations_resolve`

### CODE REVIEW

**Comment by moonbox3** [comment]
## Automated PR Review Round 1: CHANGES_REQUESTED

Issue: #1
The core fix for issue #1 looks correct: resolving postponed (`from __future__ import annotations`) handler parameter annotations via `typing.get_type_hints()` in `_validate_handler_signature` and reusing the resolved `ctx` annotation for the explicit-types decorator path addresses the string-annotation failure mode. A regression test was added and exercises the primary class-based handler scenario.

However, the PR also bundles unrelated GitHub Actions workflows that materially change the repo’s security posture, and the new type-hint resolution logic currently fails open by swallowing broad exceptions with no signal—both are blockers for merging as-is.

Reviewer summaries:
- Reviewer 1: All reviewers agree the annotation-resolution change fixes the postponed-annotations bug and the regression test targets the reported scenario.
- Reviewer 2: All reviewers flag the added GitHub Actions workflows as out-of-scope and a significant security/supply-chain risk (cloning/running external code with secrets).
- Reviewer 3: Multiple reviewers flag broad exception swallowing around `get_type_hints()` as a reliability/correctness issue that can silently reintroduce the original failure mode; one reviewer additionally notes potential perf/side-effect risks of evaluating hints at runtime.

Blocking issues:
- Remove/split the unrelated GitHub Actions workflow additions (`.github/workflows/ai-triage.yml`, `.github/workflows/docs-update.yml`). They introduce high-risk automation (cloning external/private repos using secrets, executing external code, running/pulling images) and are out of scope for an annotation bugfix; this needs separate, security-reviewed treatment.
- Make type-hint resolution failures deterministic and debuggable: `_validate_handler_signature` currently catches broad exceptions from `typing.get_type_hints()` and silently falls back to raw annotations. Narrow the exceptions to expected resolution failures and/or emit a debug/warn log (or raise an actionable error) so postponed-annotation resolution doesn’t silently regress.
- Address runtime evaluation concerns of `get_type_hints(include_extras=True)`: evaluating forward refs can have side effects and can be expensive if called repeatedly. Ensure it’s not on a hot path and consider caching per-function results (or otherwise limiting evaluation) to avoid reliability/perf issues during registration/initialization.

Suggestions:
- Add tests for the explicit-types decorator path (`@handler(input=..., output=...)`) with postponed `ctx: WorkflowContext[...]` to ensure the reused `ctx_annotation` is resolved (not a string) and behavior remains correct when `ctx` is omitted/unannotated in that mode.
- Add a negative test where type hints cannot be resolved (e.g., missing forward ref) and assert behavior is clear (either an actionable exception or a logged fallback), rather than silently proceeding to a confusing downstream error.
- Consider using `inspect.get_annotations(..., eval_str=True)` on Python 3.11+ (with a compatibility fallback) to reduce reliance on broad exception handling and to keep annotation evaluation behavior consistent.
- If nested/local definitions are supported, consider whether the `localns` used for resolution is sufficient; add a targeted test or adjust namespace handling accordingly.

**Comment by moonbox3** [comment]
## Automated PR Review Round 2: CHANGES_REQUESTED

Issue: #1
PR #5’s core change—resolving postponed/forward-ref annotations during handler signature validation via `typing.get_type_hints` (optionally with `include_extras`)—is directionally correct for issue #1, and the added regression test is a useful start. However, the PR currently broadens scope with unrelated GitHub Actions workflows that increase security/supply-chain risk, and the implementation masks type-hint resolution failures by swallowing exceptions and falling back to raw annotations. Test coverage also does not match the behaviors claimed/needed (explicit-types path, unresolvable forward refs, and failure-mode assertions). As-is, this should not be merged.

Reviewer summaries:
- Reviewer 1: Reviewer 1: Fix approach is correct, but workflows added are unrelated and risky; `get_type_hints` failures are silently ignored; tests claimed in description are missing; needs clearer failure behavior and more tests.
- Reviewer 2: Reviewer 2: Core fix is good, but new workflows introduce significant secret/supply-chain risk and should be removed/split; exception handling is too broad and hides bugs; tests don’t cover important edge cases.
- Reviewer 3: Reviewer 3: Implementation and one test are a good start, but missing negative/edge-case tests; broad exception swallowing risks silent mis-validation; current test is brittle and should better assert externally observable behavior.

Blocking issues:
- Security/scope: Remove or split out the newly added GitHub Actions workflows (`.github/workflows/ai-triage.yml`, `docs-update.yml`). They are unrelated to issue #1 and introduce material credential/supply-chain risk by cloning repos with secrets and executing unpinned external code/images.
- Correctness/diagnostics: `_validate_handler_signature` must not blanket-catch `Exception` around `typing.get_type_hints(...)` and silently fall back to `{}`/raw annotations. Narrow the exceptions and/or re-raise a clear, actionable error when annotations cannot be resolved (especially under `from __future__ import annotations`).
- Tests/coverage: Add missing tests to cover (1) unresolvable forward references producing a deterministic, actionable failure at registration/decoration time, and (2) the explicit-types decorator path where ctx annotation is postponed (and is reused/resolved consistently). Current diff only includes a basic success-path test and does not validate these behaviors.

Suggestions:
- Prefer `inspect.get_annotations(func, eval_str=True)` where available or keep `get_type_hints`, but ensure resolution failures produce a clear `ValueError` (include handler name and offending annotation) rather than silently proceeding.
- Add an additional assertion path in tests that exercises the place where resolved types are consumed (not only internal `_handlers`/`_handler_spec`), to reduce brittleness and better catch real regressions.
- If `include_extras=True` is used, add/parametrize a test for `typing.Annotated`/extras to ensure behavior is stable.
- If workflows are intended long-term, harden them in a separate PR: avoid embedding tokens in clone URLs, pin actions/images, and minimize permissions.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 1]
```diff
@@ -712,30 +710,49 @@ def _validate_handler_signature(
     signature = inspect.signature(func)
     params = list(signature.parameters.values())
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:712]: Blocking (reliability/debuggability): `get_type_hints` failures are swallowed via broad `except Exception` and then fall back to raw annotations; this can mask real resolution errors and reintroduce unresolved string annotations. Restrict exceptions you tolerate and/or raise a targeted ValueError that preserves the original exception context.

File: python/packages/core/agent_framework/_workflows/_workflow_context.py [hunk 2]
```diff
@@ -146,6 +146,12 @@ def validate_workflow_context_annotation(
             f"where T is output message type and U is workflow output type"
         )
 
+    if isinstance(annotation, str):
+        raise ValueError(
+            f"{context_description} {parameter_name} annotation '{annotation}' could not be resolved. "
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_workflow_context.py:151]: Blocking (test coverage): this unresolved-string guard is a key behavior change, but there’s no test that proves it triggers (and yields the intended error) when `WorkflowContext[...]` remains a forward-ref string. Add a regression test that produces a string annotation here and asserts the `ValueError` message.

**Review by moonbox3 — COMMENTED** [review]
## Automated PR Review Round 1: CHANGES_REQUESTED

Issue: #1
The core fix—resolving postponed/forward-ref annotations via `typing.get_type_hints` before validating `WorkflowContext[...]`—looks correct and is backed by regression tests for `__future__` annotations. However, the PR also adds two GitHub Actions workflows with high-risk secret handling and supply-chain exposure unrelated to issue #1, and the current implementation silently swallows `get_type_hints` failures, risking hard-to-debug fallbacks. Request changes to remove/split the workflows and to make type-hint resolution failures deterministic and well-tested.

Reviewer summaries:
- Reviewer 1: Bugfix and regression tests are good, but new GitHub workflows are a high-risk unrelated addition and `get_type_hints` errors are swallowed.
- Reviewer 2: Directionally correct typing fix, but workflows use privileged PATs and run external code; also exception handling hides type resolution failures.
- Reviewer 3: Test coverage is missing for new guard/fallback paths, leaving risk around silent misconfiguration and regressions.

Blocking issues:
- Remove or split the newly added GitHub Actions workflows from this PR: they clone and execute code from another repository using privileged secrets/PATs (including tokens embedded in clone URLs) and are unrelated to the handler-annotation fix, creating a significant supply-chain and secrets-exposure risk.
- In `python/packages/core/agent_framework/_workflows/_executor.py` `_validate_handler_signature`, do not swallow broad `Exception` from `typing.get_type_hints` without surfacing the root cause; restrict to expected resolution errors and/or raise a clearer validation error so unresolved annotations don’t silently fall back and fail later with less context.
- Add tests for the newly introduced guard/failure paths: (a) unresolved forward-ref/string `WorkflowContext[...]` annotations should reliably raise a deterministic, actionable `ValueError`, and (b) behavior when `get_type_hints` fails (e.g., `NameError`/`TypeError`) should be covered so the fix cannot silently degrade.

Suggestions:
- Add a negative test asserting the exact error message when a string annotation reaches `validate_workflow_context_annotation`, to keep error clarity stable over time.
- Add coverage for explicit handler typing mode (e.g., `@handler(input=..., output=...)`) with postponed annotations to ensure ctx/return type resolution is consistent across modes.
- If workflows are reintroduced in a separate PR, redesign with least privilege (avoid token-in-URL, prefer safer auth patterns, pin actions SHAs, and restrict privileged execution with protected environments/approvals).

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:712]: Automated status: round 1. Accepted for implementation in the next remediation pass.

File: python/packages/core/agent_framework/_workflows/_workflow_context.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_workflow_context.py:151]: Automated status: round 1. Accepted for implementation in the next remediation pass.
